File: apps/cortex-ai-worker/src/cortex_ai_worker/api/routes/rag.py
"""RAG endpoints — called internally by cortex-server.

Routes:
  POST /rag/retrieve         — synchronous JSON retrieval
  POST /rag/retrieve/stream  — Server-Sent Events streaming retrieval
"""

from __future__ import annotations

import json
import time
from collections.abc import AsyncGenerator
from typing import Any

import structlog
from fastapi import APIRouter, Depends, Header, HTTPException
from fastapi.responses import StreamingResponse

from ...config import settings
from ...rag.pipeline import RAGPipeline, RetrieveRequest, RetrieveResponse

logger = structlog.get_logger(__name__)

router = APIRouter()

# ─── Auth dependency ───────────────────────────────────────────────────────────

_WORKER_SECRET_HEADER = "x-worker-secret"  # noqa: S105 — header name, not a credential


async def require_worker_secret(
    x_worker_secret: str | None = Header(default=None, alias=_WORKER_SECRET_HEADER),
) -> None:
    """Validate X-Worker-Secret header.  Returns 403 if missing or incorrect."""
    if x_worker_secret is None or x_worker_secret != settings.worker_secret:
        raise HTTPException(status_code=403, detail="Invalid or missing X-Worker-Secret")


# ─── Shared pipeline instance ─────────────────────────────────────────────────

_pipeline = RAGPipeline()


# ─── POST /rag/retrieve ───────────────────────────────────────────────────────


@router.post(
    "/retrieve",
    response_model=RetrieveResponse,
    dependencies=[Depends(require_worker_secret)],
    summary="RAG retrieve — synchronous JSON",
)
async def rag_retrieve(request: RetrieveRequest) -> RetrieveResponse:
    """Run RAG retrieval and return results as JSON.

    Auth: X-Worker-Secret header required.
    """
    logger.info(
        "rag_retrieve_request",
        content_types=request.content_types,
        top_k=request.top_k,
        rerank=request.rerank,
        has_acl=bool(request.allowed_entry_ids),
    )

    response = await _pipeline.retrieve(request)

    logger.info(
        "rag_retrieve_complete",
        total=response.total,
        latency_ms=response.latency_ms,
        reranked=response.reranked,
    )

    return response


# ─── POST /rag/retrieve/stream ────────────────────────────────────────────────


@router.post(
    "/retrieve/stream",
    dependencies=[Depends(require_worker_secret)],
    summary="RAG retrieve — SSE streaming",
)
async def rag_retrieve_stream(request: RetrieveRequest) -> StreamingResponse:
    """Run RAG retrieval and stream results as Server-Sent Events.

    SSE event types (in order):
      event: chunk
      data: { chunk_index, content_entry_id, content_type, chunk_text, score, freshness_score }

      event: done
      data: { total, latency_ms, reranked }

    Auth: X-Worker-Secret header required.
    """

    async def _event_stream() -> AsyncGenerator[str, None]:
        start = time.monotonic()
        try:
            response = await _pipeline.retrieve(request)
            for chunk in response.chunks:
                payload: dict[str, Any] = {
                    "chunk_index": chunk.chunk_index,
                    "content_entry_id": chunk.content_entry_id,
                    "content_type": chunk.content_type,
                    "chunk_text": chunk.chunk_text,
                    "score": chunk.score,
                    "freshness_score": chunk.freshness_score,
                    "published_at": chunk.published_at,
                    "metadata": chunk.metadata,
                }
                yield f"event: chunk\ndata: {json.dumps(payload)}\n\n"

            done_payload: dict[str, Any] = {
                "total": response.total,
                "latency_ms": int((time.monotonic() - start) * 1000),
                "reranked": response.reranked,
            }
            yield f"event: done\ndata: {json.dumps(done_payload)}\n\n"

        except Exception as exc:
            error_payload: dict[str, Any] = {"error": str(exc), "type": type(exc).__name__}
            yield f"event: error\ndata: {json.dumps(error_payload)}\n\n"
            logger.error("rag_stream_error", error=str(exc), error_type=type(exc).__name__)

    return StreamingResponse(
        _event_stream(),
        media_type="text/event-stream",
        headers={
            "Cache-Control": "no-cache",
            "X-Accel-Buffering": "no",
            "Connection": "keep-alive",
        },
    )

File: apps/cortex-server/src/routes/rag.ts
/**
 * RAG REST API routes.
 *
 * POST /api/rag/retrieve        — synchronous JSON retrieval
 * POST /api/rag/retrieve/stream — SSE streaming retrieval
 *
 * Both require authentication (session cookie or Bearer API key).
 */

import type { FastifyInstance, FastifyPluginAsync } from "fastify";
import fp from "fastify-plugin";
import { requireAuth } from "../modules/rbac/rbac.middleware.js";
import { ragService } from "../modules/rag/rag.service.js";
import type { AuthenticatedUser } from "../modules/auth/auth.service.js";
import type { RagFormat } from "../modules/rag/rag.service.js";

// ─── Body types ───────────────────────────────────────────────────────────────

interface RagRetrieveBody {
    query: string;
    contentTypes?: string[];
    topK?: number;
    rerank?: boolean;
    format?: RagFormat;
    maxTokens?: number;
}

interface RagStreamBody {
    query: string;
    contentTypes?: string[];
    topK?: number;
}

// ─── Plugin ───────────────────────────────────────────────────────────────────

const ragRoutes: FastifyPluginAsync = async (app: FastifyInstance) => {
    // POST /api/rag/retrieve — synchronous JSON
    app.post<{ Body: RagRetrieveBody }>(
        "/api/rag/retrieve",
        {
            preHandler: requireAuth(),
            schema: {
                tags: ["RAG"],
                summary: "Retrieve semantically relevant chunks for a query",
                body: {
                    type: "object",
                    required: ["query"],
                    properties: {
                        query: { type: "string", minLength: 1, maxLength: 1000 },
                        contentTypes: {
                            type: "array",
                            items: { type: "string" },
                        },
                        topK: { type: "integer", minimum: 1, maximum: 50, default: 10 },
                        rerank: { type: "boolean", default: false },
                        format: {
                            type: "string",
                            enum: ["chunks", "context", "json"],
                            default: "chunks",
                        },
                        maxTokens: { type: "integer", minimum: 100, maximum: 32000, default: 4000 },
                    },
                },
            },
        },
        async (request, reply) => {
            try {
                const user = request.user as AuthenticatedUser;
                const body = request.body;

                const result = await ragService.retrieve({
                    query: body.query,
                    contentTypes: body.contentTypes,
                    topK: body.topK,
                    rerank: body.rerank,
                    format: body.format,
                    maxTokens: body.maxTokens,
                    user,
                });

                return { data: result };
            } catch (err) {
                const message = err instanceof Error ? err.message : "Unknown error";
                request.log.error({ msg: "rag_retrieve_error", error: message });

                if (message.includes("required") || message.includes("exceeds maximum")) {
                    return reply.status(400).send({
                        error: { code: "BAD_REQUEST", message },
                    });
                }

                return reply.status(500).send({
                    error: { code: "INTERNAL_SERVER_ERROR", message: "RAG retrieval failed" },
                });
            }
        }
    );

    // POST /api/rag/retrieve/stream — SSE streaming
    app.post<{ Body: RagStreamBody }>(
        "/api/rag/retrieve/stream",
        {
            preHandler: requireAuth(),
            schema: {
                tags: ["RAG"],
                summary: "Stream RAG retrieval as Server-Sent Events",
                body: {
                    type: "object",
                    required: ["query"],
                    properties: {
                        query: { type: "string", minLength: 1, maxLength: 1000 },
                        contentTypes: {
                            type: "array",
                            items: { type: "string" },
                        },
                        topK: { type: "integer", minimum: 1, maximum: 50, default: 10 },
                    },
                },
            },
        },
        async (request, reply) => {
            const user = request.user as AuthenticatedUser;
            const body = request.body;

            // Set SSE headers before streaming
            void reply.raw.writeHead(200, {
                "Content-Type": "text/event-stream",
                "Cache-Control": "no-cache",
                "X-Accel-Buffering": "no",
                Connection: "keep-alive",
            });

            try {
                const generator = ragService.retrieveStream({
                    query: body.query,
                    contentTypes: body.contentTypes,
                    topK: body.topK,
                    user,
                });

                for await (const event of generator) {
                    if (event.type === "chunk") {
                        const payload = JSON.stringify({
                            chunk_index: event.chunkIndex,
                            content_entry_id: event.contentEntryId,
                            content_type: event.contentType,
                            chunk_text: event.chunkText,
                            score: event.score,
                            freshness_score: event.freshnessScore,
                            published_at: event.publishedAt,
                            metadata: event.metadata,
                        });
                        reply.raw.write(`event: chunk\ndata: ${payload}\n\n`);
                    } else if (event.type === "done") {
                        const payload = JSON.stringify({
                            total: event.total,
                            latency_ms: event.latencyMs,
                            reranked: event.reranked,
                        });
                        reply.raw.write(`event: done\ndata: ${payload}\n\n`);
                    } else if (event.type === "error") {
                        const payload = JSON.stringify({ error: event.error });
                        reply.raw.write(`event: error\ndata: ${payload}\n\n`);
                    }
                }
            } catch (err) {
                const message = err instanceof Error ? err.message : "Unknown error";
                request.log.error({ msg: "rag_stream_error", error: message });
                reply.raw.write(
                    `event: error\ndata: ${JSON.stringify({ error: message })}\n\n`
                );
            } finally {
                reply.raw.end();
            }
        }
    );
};

export default fp(ragRoutes, { name: "rag-routes" });
